materialize/lifecycle.go

```diff
@@ -488,28 +538,100 @@ func (it *StoreIterator) Commit() *TransactionRequest_Commit {
 // Transactor is a store-agnostic interface for a materialization driver
 // that implements Flow materialization protocol transactions.
 type Transactor interface {
-	// Load implements the transaction "load" phase by:
-	// * Consuming Load requests from the LoadIterator.
-	// * Awaiting a Prepare request or |commitCh| before reading from the store.
-	// * Invoking loaded() with loaded documents.
+	// Load implements the transaction load phase by consuming Load requests
+	// from the LoadIterator until a Prepare request is read. Requested keys
+	// are not known to exist in the store, and very often they won't.
+	// Load can ignore keys which are not found in the store. Before Load returns,
+	// though, it must ensure loaded() is called for all found documents.
 	//
 	// Loads of transaction T+1 will be invoked after Store of T+0, but
-	// concurrently with Commit of T+0. This is an important optimization that
-	// allows the driver to immediately begin the work of T+1, for instance by
-	// staging Load keys into a temporary table which will later be joined and
-	// queried upon a future Prepare.
+	// concurrently with the commit and acknowledgement of T+0. This is an
+	// important optimization that allows the driver to immediately begin the work
+	// of T+1, for instance by staging Load keys into a temporary table which will
+	// be joined and queried once the LoadIterator is drained.
 	//
 	// But, the driver contract is that documents loaded in T+1 must reflect
-	// stores of T+0 (a.k.a. "read committed"). The driver must therefore await
-	// either a Prepare or |commitCh| before reading from the store to ensure
-	// this contract is met.
-	Load(_ *LoadIterator, commitCh <-chan struct{}, loaded func(binding int, doc json.RawMessage) error) error
-	// Prepare begins the transaction "store" phase.
-	Prepare(*TransactionRequest_Prepare) (*TransactionResponse_Prepared, error)
+	// stores of T+0 (a.k.a. "read committed"). The driver must ensure that the
+	// prior transaction is fully reflected in the store before attempting to
+	// load any keys.
+	//
+	// Load is given two channels, priorCommittedCh and priorAcknowledgedCh,
+	// to help it understand where the prior transaction is in its lifecycle.
+	//  - priorCommittedCh selects (is closed) upon the completion of Transactor.Commit.
+	//  - priorAcknowledgedCh selects (is closed) upon completion of Transactor.Acknowledge.
+	//
+	// * In the simple case where the driver uses implementation pattern
+	//   "Recovery Log with Non-Transactional Store", the driver is free
+	//   to simply ignore these signals. It may issue lookups to the store and call
+	//   back to loaded() when those lookups succeed.
```

Comment:
I really appreciate referencing the Implementation Patterns here. There are a lot of subtle interactions going on, and a lot of simple ways to use these pieces and parts wrong.